src/app/OperationalDeviceProxy.cpp

```diff
@@ -222,9 +221,9 @@ void OperationalDeviceProxy::HandleCASEConnectionFailure(void * context, CASECli
 
     device->mState = State::Initialized;
 
+    // Device failure callbacks are expected to cleanup any CASE sessions.
     device->DequeueConnectionSuccessCallbacks(/* executeCallback */ false);
     device->DequeueConnectionFailureCallbacks(error, /* executeCallback */ true);
```

Comment:
How are callbacks supposed to get this right, given that there can be multiple callbacks multiplexed on the same device?  The real issue here is that some failure callbacks destroy this object.  So what we want is this:
```suggestion
    device->CloseCASESession();
    device->DequeueConnectionSuccessCallbacks(/* executeCallback */ false);
    device->DequeueConnectionFailureCallbacks(error, /* executeCallback */ true);
    // Do not touch device anymore; it might have been destroyed by a failure callback.
```

and a similar change in the success case, actually.